File: BE/src/main/java/com/example/WebApartment/Service/ChatService.java
package com.example.WebApartment.Service;

import com.example.WebApartment.DTO.*;
import com.example.WebApartment.Models.*;
import com.example.WebApartment.Repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ChatService {

    private final PhongChatRepository phongChatRepository;
    private final TinNhanRepository tinNhanRepository;
    private final NguoiDungRepository nguoiDungRepository;
    private final BaiDangRepository baiDangRepository;

    @Transactional
    public ChatRoomDTO getOrCreateRoom(CreateChatRoomRequest request) {
        String user1 = request.getMaNguoiDung1();
        String user2 = request.getMaNguoiDung2();
        String loai = request.getLoaiPhongChat();

        PhongChat existing = phongChatRepository
                .findByNguoiDung1_MaNguoiDungAndNguoiDung2_MaNguoiDungAndLoaiPhongChat(user1, user2, loai)
                .or(() -> phongChatRepository
                        .findByNguoiDung2_MaNguoiDungAndNguoiDung1_MaNguoiDungAndLoaiPhongChat(user1, user2, loai))
                .orElse(null);

        if (existing != null) {
            return toRoomDTO(existing);
        }

        NguoiDung nguoiDung1 = nguoiDungRepository.findById(user1)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy người dùng 1"));

        NguoiDung nguoiDung2 = nguoiDungRepository.findById(user2)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy người dùng 2"));

        BaiDang baiDang = null;
        if (request.getMaBaiDang() != null && !request.getMaBaiDang().isBlank()) {
            baiDang = baiDangRepository.findById(request.getMaBaiDang())
                    .orElseThrow(() -> new RuntimeException("Không tìm thấy bài đăng"));
        }

        LocalDateTime now = LocalDateTime.now();

        PhongChat room = PhongChat.builder()
                .maPhongChat(generateId("PC"))
                .loaiPhongChat(loai)
                .nguoiDung1(nguoiDung1)
                .nguoiDung2(nguoiDung2)
                .baiDang(baiDang)
                .tinNhanCuoi("")
                .thoiGianTinNhanCuoi(now)
                .ngayTao(now)
                .build();

        return toRoomDTO(phongChatRepository.save(room));
    }

    public List<ChatRoomDTO> getRoomsByUser(String maNguoiDung) {
        return phongChatRepository
                .findByNguoiDung1_MaNguoiDungOrNguoiDung2_MaNguoiDungOrderByThoiGianTinNhanCuoiDesc(
                        maNguoiDung,
                        maNguoiDung
                )
                .stream()
                .map(this::toRoomDTO)
                .toList();
    }

    public List<ChatMessageDTO> getMessages(String maPhongChat) {
        return tinNhanRepository
                .findByPhongChat_MaPhongChatOrderByThoiGianGuiAsc(maPhongChat)
                .stream()
                .map(this::toMessageDTO)
                .toList();
    }

    @Transactional
    public ChatMessageDTO sendMessage(SendMessageRequest request) {
        PhongChat room = phongChatRepository.findById(request.getMaPhongChat())
                .orElseThrow(() -> new RuntimeException("Không tìm thấy phòng chat"));

        NguoiDung sender = nguoiDungRepository.findById(request.getMaNguoiGui())
                .orElseThrow(() -> new RuntimeException("Không tìm thấy người gửi"));

        LocalDateTime now = LocalDateTime.now();

        TinNhan message = TinNhan.builder()
                .maTinNhan(generateId("TN"))
                .phongChat(room)
                .nguoiGui(sender)
                .noiDung(request.getNoiDung())
                .loaiTinNhan(request.getLoaiTinNhan() == null ? "TEXT" : request.getLoaiTinNhan())
                .tepDinhKemUrl(request.getTepDinhKemUrl())
                .trangThai("SENT")
                .thoiGianGui(now)
                .build();

        TinNhan saved = tinNhanRepository.save(message);

        room.setTinNhanCuoi(request.getNoiDung());
        room.setThoiGianTinNhanCuoi(now);
        phongChatRepository.save(room);

        return toMessageDTO(saved);
    }

    private ChatRoomDTO toRoomDTO(PhongChat room) {
        return ChatRoomDTO.builder()
                .maPhongChat(room.getMaPhongChat())
                .loaiPhongChat(room.getLoaiPhongChat())
                .maNguoiDung1(room.getNguoiDung1() != null ? room.getNguoiDung1().getMaNguoiDung() : null)
                .tenNguoiDung1(room.getNguoiDung1() != null ? room.getNguoiDung1().getHoVaTen() : null)
                .maNguoiDung2(room.getNguoiDung2() != null ? room.getNguoiDung2().getMaNguoiDung() : null)
                .tenNguoiDung2(room.getNguoiDung2() != null ? room.getNguoiDung2().getHoVaTen() : null)
                .maBaiDang(room.getBaiDang() != null ? room.getBaiDang().getMaBaiDang() : null)
                .tieuDeBaiDang(room.getBaiDang() != null ? room.getBaiDang().getTieuDe() : null)
                .tinNhanCuoi(room.getTinNhanCuoi())
                .thoiGianTinNhanCuoi(room.getThoiGianTinNhanCuoi())
                .ngayTao(room.getNgayTao())
                .build();
    }

    private ChatMessageDTO toMessageDTO(TinNhan message) {
        return ChatMessageDTO.builder()
                .maTinNhan(message.getMaTinNhan())
                .maPhongChat(message.getPhongChat() != null ? message.getPhongChat().getMaPhongChat() : null)
                .maNguoiGui(message.getNguoiGui() != null ? message.getNguoiGui().getMaNguoiDung() : null)
                .tenNguoiGui(message.getNguoiGui() != null ? message.getNguoiGui().getHoVaTen() : null)
                .noiDung(message.getNoiDung())
                .loaiTinNhan(message.getLoaiTinNhan())
                .tepDinhKemUrl(message.getTepDinhKemUrl())
                .trangThai(message.getTrangThai())
                .thoiGianGui(message.getThoiGianGui())
                .build();
    }

    private String generateId(String prefix) {
        return prefix + UUID.randomUUID().toString()
                .replace("-", "")
                .substring(0, 10)
                .toUpperCase();
    }
}

File: FE/src/services/api/ChatService.ts
import axiosClient from './AxiosClient';

export interface ChatRoomDTO {
  maPhongChat: string;
  loaiPhongChat: string;
  maNguoiDung1: string;
  tenNguoiDung1: string;
  maNguoiDung2: string;
  tenNguoiDung2: string;
  maBaiDang?: string | null;
  tieuDeBaiDang?: string | null;
  tinNhanCuoi?: string;
  thoiGianTinNhanCuoi?: string;
  ngayTao?: string;
}

export interface ChatMessageDTO {
  maTinNhan: string;
  maPhongChat: string;
  maNguoiGui: string;
  tenNguoiGui: string;
  noiDung: string;
  loaiTinNhan: string;
  tepDinhKemUrl?: string | null;
  trangThai?: string;
  thoiGianGui: string;
}

export interface CreateChatRoomRequest {
  maNguoiDung1: string;
  maNguoiDung2: string;
  maBaiDang?: string | null;
  loaiPhongChat: string;
}

export interface SendMessageRequest {
  maPhongChat: string;
  maNguoiGui: string;
  noiDung: string;
  loaiTinNhan?: string;
  tepDinhKemUrl?: string | null;
}

export interface ChatAttachmentDTO {
  url: string;
  originalName?: string;
  contentType?: string;
  size?: number;
}

/**
 * Lấy hoặc tạo phòng chat giữa 2 người dùng
 */
export const getOrCreateRoom = async (request: CreateChatRoomRequest): Promise<ChatRoomDTO> => {
  const response = await axiosClient.post<ChatRoomDTO>('/api/v1/chat/rooms', request);
  return response.data;
};

/**
 * Lấy danh sách tất cả các phòng chat của một người dùng
 */
export const getRoomsByUser = async (maNguoiDung: string): Promise<ChatRoomDTO[]> => {
  const response = await axiosClient.get<ChatRoomDTO[]>(`/api/v1/chat/rooms/user/${maNguoiDung}`);
  return response.data;
};

/**
 * Lấy lịch sử tin nhắn của một phòng chat cụ thể
 */
export const getMessages = async (maPhongChat: string): Promise<ChatMessageDTO[]> => {
  const response = await axiosClient.get<ChatMessageDTO[]>(`/api/v1/chat/rooms/${maPhongChat}/messages`);
  return response.data;
};

/**
 * Gửi tin nhắn qua REST API (fallback hoặc đính kèm ảnh)
 */
export const sendMessageRest = async (request: SendMessageRequest): Promise<ChatMessageDTO> => {
  const response = await axiosClient.post<ChatMessageDTO>('/api/v1/chat/messages', request);
  return response.data;
};

/**
 * Upload ảnh chat từ thiết bị, trả về URL đã lưu trên Cloudinary
 */
export const uploadChatImage = async (file: File): Promise<ChatAttachmentDTO> => {
  const formData = new FormData();
  formData.append('file', file);

  const response = await axiosClient.post<ChatAttachmentDTO>(
    '/api/v1/chat/attachments',
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );

  return response.data;
};
